File: TwoPointers/RainTraps.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int traps = 0;
        int start = 0;
        int end = height.size() - 1; 
        int maxStart = height[start];
        int maxEnd = height[end];
        while(start < end)
        {
            if(maxStart < maxEnd)
            {
                start++;
                if(min(maxStart,maxEnd) - height[start] > 0)
                    traps += min(maxStart,maxEnd) - height[start];
                maxStart = max(maxStart,height[start]);
            }
            else
            {
                end--;
                if(min(maxStart,maxEnd) - height[end] > 0)
                    traps += min(maxStart,maxEnd) - height[end];
                maxEnd = max(maxEnd,height[end]);
            }
        }
        return traps;
    }
};

File: TwoPointers/RainTraps.py
class Solution(object):
    def trap(self, height):
        start, end = 0, len(height)-1
        maxS, maxE = height[0],height[len(height)-1]
        traps = 0
        while start < end:
            if maxS < maxE:
                start += 1
                if maxS - height[start] > 0:
                    traps += maxS - height[start]
                maxS = max(maxS,height[start])
            else:
                end -= 1
                if maxE - height[end] > 0:
                    traps += maxE - height[end]
                maxE = max(maxE,height[end])
        return traps